typescript/utils/src/logging.ts

```diff
@@ -1,47 +1,36 @@
-import { safelyAccessEnvVar } from './env';
-
-/* eslint-disable no-console */
-type LOG_LEVEL = 'none' | 'error' | 'warn' | 'info' | 'debug' | 'trace';
-
-const ENV_LOG_LEVEL = (
-  safelyAccessEnvVar('LOG_LEVEL') ?? 'debug'
-).toLowerCase() as LOG_LEVEL;
-const LOG_TRACE = ENV_LOG_LEVEL == 'trace';
-const LOG_DEBUG = LOG_TRACE || ENV_LOG_LEVEL == 'debug';
-const LOG_INFO = LOG_DEBUG || ENV_LOG_LEVEL == 'info';
-const LOG_WARN = LOG_INFO || ENV_LOG_LEVEL == 'warn';
-const LOG_ERROR = LOG_WARN || ENV_LOG_LEVEL == 'error';
-
-export function trace(message: string, data?: any) {
-  if (LOG_TRACE) logWithFunction(console.trace, 'trace', message, data);
-}
-
-export function debug(message: string, data?: any) {
-  if (LOG_DEBUG) logWithFunction(console.debug, 'debug', message, data);
-}
-
-export function log(message: string, data?: any) {
-  if (LOG_INFO) logWithFunction(console.log, 'info', message, data);
-}
-
-export function warn(message: string, data?: any) {
-  if (LOG_WARN) logWithFunction(console.warn, 'warn', message, data);
-}
-
-export function error(message: string, data?: any) {
-  if (LOG_ERROR) logWithFunction(console.error, 'error', message, data);
-}
-
-function logWithFunction(
-  logFn: (...contents: any[]) => void,
-  level: LOG_LEVEL,
-  message: string,
-  data?: any,
-) {
-  const fullLog = {
-    ...data,
-    level,
-    message,
-  };
-  logFn(JSON.stringify(fullLog));
-}
+import { LevelWithSilent, pino } from 'pino';
+
+import { envVarToBoolean, safelyAccessEnvVar } from './env';
+
+const DEFAULT_LOG_LEVEL = 'warn';
+
+let ENV_LOG_LEVEL = (
+  safelyAccessEnvVar('LOG_LEVEL') ?? DEFAULT_LOG_LEVEL
+).toLowerCase() as LevelWithSilent | 'none' | 'off';
+// For backwards compat and also to match agent level options
+if (ENV_LOG_LEVEL === 'none' || ENV_LOG_LEVEL === 'off')
+  ENV_LOG_LEVEL = 'silent';
+
+const ENV_LOG_PRETTY = envVarToBoolean(safelyAccessEnvVar('LOG_PRETTY'));
+
+export const rootLogger = pino({
+  level: ENV_LOG_LEVEL,
+  name: 'hyperlane',
+  formatters: {
+    // Remove pino's default bindings of hostname and pid
+    bindings: () => ({}),
+  },
+  hooks: {
+    logMethod(inputArgs, method, level) {
+      // Pino has no simple way of setting custom log shapes and they
+      // recommend against using pino-pretty in production so when
+      // pretty is enabled we circumvent pino and log directly to console
+      if (ENV_LOG_PRETTY && level >= pino.levels.values[ENV_LOG_LEVEL]) {
+        // eslint-disable-next-line no-console
+        console.log(inputArgs[0]);
+        return null;
```

Comment:
I dont really understand this

Reply:
Does the comment above help? We're constrained by what pino lets us do. Short of restarting this work with a different lib, I'm not sure we can do better